### PR TITLE
Simple ValueTuple implementation ...

### PR DESCRIPTION
**Pull request by DaEgi01**
... meant as a replacement for IntVector2 and friends.

### CODE REVIEW

**Comment by kvakvs** [comment]
Not sure what problem this solves. We do not have any other use for this than vector of width and height.

File: TLM/TLM/ValueTuple.cs [hunk 1]
```diff
@@ -0,0 +1,44 @@
+using System;
+using System.Collections.Generic;
+using System.Runtime.InteropServices;
+
+namespace TrafficManager {
```
Comment (krzychu124) [inline TLM/TLM/ValueTuple.cs:5]: Hmm why not in `\Util`?

**Review by krzychu124 — CHANGES_REQUESTED** [review]
See comment

File: TLM/TLM/ValueTuple.cs (hunk 1 above)
Comment (DaEgi01) [reply TLM/TLM/ValueTuple.cs:5]: Because its so basic/core it could pollute all classes with \Util.
In .net its under the System namespace.

**Review by krzychu124 — APPROVED** [review]

**Comment by DaEgi01** [comment]
> Not sure what problem this solves.
The same problem that IntVector2 solve, only in a more generic and standardized way.

> We do not have any other use for this than vector of width and height.
But we do? I assume you are not a tuple user, otherwise you'd see a lot of potential here and there.
Once its in place, there is no need anymore to work around the issue that you can't pass or return multiple values easily.
See AdvancedParkingManager.GetBuildingInfoViewColor for example, or ExtCitizenInstanceManager.FindPathPosition.
I mean common, posA, posB, distA, distB just screams "I don't have tuples" into my face at least, out parameters too.
If we had tuples from day one, I'd argue there would be no out parameters at all in c#.

> I can approve and sure go ahead if this makes you more happy.
It does a lot :)

> But there is no use for a generic pair 😀
Tuples are not limited to pairs. We can extend them easily to 3 and 4 items.
"No use" is too strong imo.
Sure you can work around it if you don't have them, but imo code comprehension and coding convenience will suffer from that.
The is also "no use" for booleans in c. But I'd still argue that every language with booleans is better of with them.
A generic concept to combine two or more values into one unit is worth it imo since it comes up a lot in programming.

**Comment by kvakvs** [comment]
No worries I know where tuple can be used, I asked in chat when i was adding this class, and apparently tuples were not in the .NET framework we are using, so i assumed they are not available.

There is a naming problem. ValueTuple is in fact a Tuple2 or ValueTuple2 (i'd like shorter name though).
There can be other size tuples, be ready to add them as necessary, Tuple3, Tuple4 and so on.

**Comment by DaEgi01** [comment]
> No worries I know where tuple can be used, I asked in chat when i was adding this class, and apparently tuples were not in the .NET framework we are using, so i assumed they are not available.
> 
> There is a naming problem. ValueTuple is in fact a Tuple2 or ValueTuple2 (i'd like shorter name though).
> There can be other size tuples, be ready to add them as necessary, Tuple3, Tuple4 and so on.

You don't need ValueTuple2, ValueTuple3, ValueTuple4 since the amount of type parameters implicitly determines what size they are.
ValueTuple2 -> ValueTuple<T1, T2>
ValueTuple3 -> ValueTuple<T1, T2, T3>
ValueTuple4 -> ValueTuple<T1, T2, T3, T4>

also kian found some neat stuff.
and once I start looking I found https://github.com/theraot/Theraot/ which has a lot of goodies from newer .net
lots of goodies from newer .net versions ported back to 3.5.
in the tuple case we get full support for named parameter, deconstruction etc.

**Pull request closed without merging**